guide.md

```diff
@@ -0,0 +1,324 @@
+
+```bazel
+# download the http_archive rule itself
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# download rules_nixpkgs
+http_archive(
+    name = "io_tweag_rules_nixpkgs",
+    strip_prefix = "rules_nixpkgs-9f08fb2322050991dead17c8d10d453650cf92b7",
+    urls = ["https://github.com/tweag/rules_nixpkgs/archive/9f08fb2322050991dead17c8d10d453650cf92b7.tar.gz"],
+    sha256 = "46aa0ca80b77848492aa1564e9201de9ed79588ca1284f8a4f76deb7a0eeccb9",
+)
+
+# load everything that rules_nixpkgs rules need to work
+load("@io_tweag_rules_nixpkgs//nixpkgs:repositories.bzl", "rules_nixpkgs_dependencies")
+rules_nixpkgs_dependencies()
+```
+
+Now we can select a version of the Nixpkgs repository from which all of our
+packages will be drawn. Any Git commit hash or tag is allowed. This way, we
+select a consistent set of package revisions in one go instead of having to
+maintain dependency versions for individual dependencies. Often this will be a
+tag for a stable release, which ensures that there has been especially
+stringent verification that packages in it will function correctly individually
+and play well together. Here we use the 22.05 release:
+
+```bazel
+load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl", "nixpkgs_git_repository")
+nixpkgs_git_repository(
+    name = "nixpkgs",
+    revision = "22.05",
+    sha256 = "",
+)
+```
+
+# A hermetic CC toolchain
+
+Now we can now define a CC toolchain that will be provisioned by Nix using
+package definitions from the 22.05 release of Nixpkgs.
+
+```bazel
+load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl", "nixpkgs_cc_configure")
+nixpkgs_cc_configure(
+  repository = "@nixpkgs",
+  name = "nixpkgs_config_cc",
+)
+```
+
+This will select the default `stdenv.cc` package as the compiler, drawing
+auxiliary tools from `stdenv.cc.bintools`. However, one can write a little bit of
+Nix to configure this:
+
+```bazel
+load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl", "nixpkgs_cc_configure")
+nixpkgs_cc_configure(
+  repository = "@nixpkgs",
+  nix_file_content = "(import <nixpkgs> {}).gcc11",
+  name = "nixpkgs_config_cc",
+)
+```
+
+`(import <nixpkgs> {})` loads the entire package set into an attribute set
+(Nix's equivalent of a record), and `.gcc11` selects one value from it. It is
+possible to customize the compiler further using more Nix if desired: you could,
+for instance, point Nix to a custom fork of Clang containing support for a new
+architecture, and have everything work seamlessly.
+
+With this, we now have a fully hermetic CC toolchain that will behave correctly
+across platforms. It remains to tell Bazel to use it by default, instead of
+`@local_config_cc`. To do so, we put the following into `.bazelrc` at the root
+of the repository:
+
+```
+build --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+run   --host_platform=@io_tweag_rules_nixpkgs//nixpkgs/platforms:host
+build --crosstool_top=@nixpkgs_config_cc//:toolchain
```

Comment:
Maybe worth mentioning [`https://github.com/bazelbuild/bazel/issues/7260`](https://github.com/bazelbuild/bazel/issues/7260) as well?

Reply:
I can't seem to get `--incompatible_enable_cc_toolchain_resolution` to work on this repo, even if I specify the platform name manually on the command line, presumably because I haven't told it to use the `@nixpkgs_config_cc` toolchain. Renaming it to `@local_config_cc` works (and indeed simplifies some of the guide). 

However, if I make that change, I don't know how to adapt the "Making Nix optional" section (or if it needs to be adapted at all: will toolchain resolution silently fall back to the default toolchain?)

Reply:
So far the explanations were very detailed, but here the user is left pondering what these two options really mean. Maybe add a few sentences about them, even if it is in a hand-wavy way?

Reply:
The repository I care about right now (Uber's Go Monorepo) has `--incompatible_enable_cc_toolchain_resolution` set by default; so I presume there are more repositories "in the wild" that uses it.

Perhaps we can add a hint about it, a TODO item, and at least a link to https://github.com/tweag/rules_nixpkgs/issues/289 ?